azure_functions_worker/http_v2.py

```diff
@@ -0,0 +1,244 @@
+            self._context_references[invoc_id] = AsyncContextReference()
+
+        await asyncio.sleep(0)
+        await self._context_references.get(
+            invoc_id).http_request_available_event.wait()
+        return self._pop_http_request(invoc_id)
+
+    async def await_http_response_async(self, invoc_id):
+        if invoc_id not in self._context_references:
+            raise Exception("No context reference found for invocation "
+                            f"{invoc_id}")
+        await asyncio.sleep(0)
+        await self._context_references.get(
+            invoc_id).http_response_available_event.wait()
+        return self._pop_http_response(invoc_id)
+
+    def _pop_http_request(self, invoc_id):
+        context_ref = self._context_references.get(invoc_id)
+        request = context_ref.http_request
+        if request is not None:
+            context_ref.http_request = None
+            return request
+
+        raise Exception(f"No http request found for invocation {invoc_id}")
+
+    def _pop_http_response(self, invoc_id):
+        context_ref = self._context_references.get(invoc_id)
+        response = context_ref.http_response
+        if response is not None:
+            context_ref.http_response = None
+            return response
+        raise Exception(f"No http response found for invocation {invoc_id}")
+
+
+def get_unused_tcp_port():
+    # Create a TCP socket
+    tcp_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    # Bind it to a free port provided by the OS
+    tcp_socket.bind(("", 0))
+    # Get the port number
+    port = tcp_socket.getsockname()[1]
+    # Close the socket
+    tcp_socket.close()
+    # Return the port number
+    return port
+
+
+def initialize_http_server(host_addr, **kwargs):
+    ext_base = HttpV2Registry.ext_base()
+    web_extension_mod_name = ext_base.ModuleTrackerMeta.get_module()
+    extension_module = importlib.import_module(web_extension_mod_name)
+    web_app_class = extension_module.WebApp
+    web_server_class = extension_module.WebServer
+
+    unused_port = get_unused_tcp_port()
+
+    app = web_app_class()
+    request_type = ext_base.RequestTrackerMeta.get_request_type()
+
+    @app.route
+    async def catch_all(request: request_type):  # type: ignore
+        invoc_id = request.headers.get(X_MS_INVOCATION_ID)
+        if invoc_id is None:
+            raise ValueError(f"Header {X_MS_INVOCATION_ID} not found")
+        logger.info(f'Received HTTP request for invocation {invoc_id}')
+        http_coordinator.set_http_request(invoc_id, request)
+        http_resp = \
+            await http_coordinator.await_http_response_async(invoc_id)
+
+        logger.info(f'Sending HTTP response for invocation {invoc_id}')
+        # if http_resp is an python exception, raise it
+        if isinstance(http_resp, Exception):
+            raise http_resp
+
+        return http_resp
+
+    web_server = web_server_class(host_addr, unused_port, app)
```

Comment:
Do we need to try/catch any of these lines below?

Reply:
no need, attribute errors will be raised if any vars/chained attritbues is missing and thrown up to rpc error response.